File: Assignment-2/code/part1-programs/scope-closures.js
// Section 2: the same three differences in JavaScript, shown by what this program prints. The three are name
// binding, closure capture and type checking, in that order.
//
// Run: node scope-closures.js

// Name binding. Both declarations below are known before this line runs, but var and let disagree about what the
// name holds until its own line is reached.

function readHoistedLimit() {
  return hoistedLimit;
}

function readBlockScopedLimit() {
  return blockScopedLimit;
}

console.log("name binding: var before its declaration ->", readHoistedLimit());

try {
  console.log("name binding: let before its declaration ->", readBlockScopedLimit());
} catch (accessBeforeInitialisation) {
  console.log("name binding: ReferenceError ->", accessBeforeInitialisation.message);
}

var hoistedLimit = 10;
let blockScopedLimit = 99;
console.log("name binding: after both declaration ->", readHoistedLimit(), readBlockScopedLimit());

// Closure capture. The same loop gives two different answers, decided by the loop keyword alone.

function buildIndexReportersWithVar() {
  const reporters = [];
  for (var index = 0; index < 3; index += 1) {
    reporters.push(() => index); // one function-scoped index, shared by all three, left holding 3 by the exit test
  }
  return reporters;
}

function buildIndexReportersWithLet() {
  const reporters = [];
  for (let index = 0; index < 3; index += 1) {
    reporters.push(() => index); // let gives each iteration its own binding, so each closure keeps its own index
  }
  return reporters;
}

console.log("closure capture: var ->", buildIndexReportersWithVar().map((reporter) => reporter()));
console.log("closure capture: let ->", buildIndexReportersWithLet().map((reporter) => reporter()));

// Type checking. No error surfaces at all. The operator decides which way to convert, and both lines run.

console.log('type checking: "1" + 1 ->', "1" + 1, "of type", typeof ("1" + 1));
console.log('type checking: "1" - 1 ->', "1" - 1, "of type", typeof ("1" - 1));

File: Assignment-2/code/part1-programs/scope-closures.py
"""Section 2: three semantic differences between Python, JavaScript and C++, shown by what this program prints.

The three are name binding, closure capture and type checking, in that order.

Run: python3 scope-closures.py
"""

# Name binding. Python resolves a global name on every call, not when the function is defined.


def report_measurement_limit():
    # measurement_limit does not exist yet. Python accepts the function anyway and looks the name up on each call.
    return measurement_limit


try:
    print("name binding: before the name exists ->", report_measurement_limit())
except NameError as missing_name:
    print("name binding: NameError ->", missing_name)

measurement_limit = 10
print("name binding: after the name exists ->", report_measurement_limit())

measurement_limit = 99
print("name binding: same function, new answer ->", report_measurement_limit())


# Closure capture. Python closes over the variable itself, not over the value it held at the time.


def build_index_reporters():
    reporters = []
    for index in range(3):
        reporters.append(lambda: index)  # all three lambdas share the one loop variable, which ends holding 2
    return reporters


def build_index_reporters_by_value():
    reporters = []
    for index in range(3):
        reporters.append(lambda captured_index=index: captured_index)  # a default argument copies the value now
    return reporters


print("closure capture: shared loop variable ->", [reporter() for reporter in build_index_reporters()])
print("closure capture: value copied per lambda ->", [reporter() for reporter in build_index_reporters_by_value()])


# Type checking. Both parts above have already printed, because nothing checked this line until the interpreter
# reached it.

try:
    print("type checking: \"1\" + 1 ->", "1" + 1)
except TypeError as mismatched_operands:
    print("type checking: TypeError ->", mismatched_operands)

print("type checking: with the conversion written out ->", "1" + str(1))
